core/src/main/scala/com/dwolla/security/crypto/Decrypt.scala

```diff
@@ -0,0 +1,158 @@
+package com.dwolla.security.crypto
+
+import cats.effect.*
+import cats.syntax.all.*
+import com.dwolla.security.crypto.DecryptToInputStream.*
+import fs2.*
+import fs2.io.{readInputStream, toInputStream}
+import org.bouncycastle.openpgp.*
+import org.bouncycastle.openpgp.operator.jcajce.*
+import org.typelevel.log4cats.{Logger, LoggerFactory}
+
+import java.io.*
+import scala.annotation.nowarn
+
+trait Decrypt[F[_]] {
+  def decrypt(key: PGPPrivateKey,
+              chunkSize: ChunkSize,
+             ): Pipe[F, Byte, Byte]
+
+  def decrypt(keyring: PGPSecretKeyRing,
+              passphrase: Array[Char],
+              chunkSize: ChunkSize,
+             ): Pipe[F, Byte, Byte]
+
+  def decrypt(keyring: PGPSecretKeyRingCollection,
+              passphrase: Array[Char],
+              chunkSize: ChunkSize,
+             ): Pipe[F, Byte, Byte]
+
+  /* the rest of these definitions just provide default values for arguments */
+  final def decrypt(key: PGPPrivateKey): Pipe[F, Byte, Byte] =
+    decrypt(key, defaultChunkSize)
+
+  final def decrypt(keyring: PGPSecretKeyRing): Pipe[F, Byte, Byte] =
+    decrypt(keyring, Array.empty[Char], defaultChunkSize)
+
+  final def decrypt(keyring: PGPSecretKeyRing, passphrase: Array[Char]): Pipe[F, Byte, Byte] =
+    decrypt(keyring, passphrase, defaultChunkSize)
+
+  final def decrypt(keyring: PGPSecretKeyRing, chunkSize: ChunkSize): Pipe[F, Byte, Byte] =
+    decrypt(keyring, Array.empty[Char], chunkSize)
+
+  final def decrypt(keyring: PGPSecretKeyRingCollection): Pipe[F, Byte, Byte] =
+    decrypt(keyring, Array.empty[Char], defaultChunkSize)
+
+  final def decrypt(keyring: PGPSecretKeyRingCollection, passphrase: Array[Char]): Pipe[F, Byte, Byte] =
+    decrypt(keyring, passphrase, defaultChunkSize)
+
+  final def decrypt(keyring: PGPSecretKeyRingCollection, chunkSize: ChunkSize): Pipe[F, Byte, Byte] =
+    decrypt(keyring, Array.empty[Char], chunkSize)
+
+}
+
+object Decrypt {
+  @nowarn("""msg=parameter (?:value )?ev in method apply is never used""")
+  def apply[F[_] : Async : Logger : LoggerFactory](implicit ev: BouncyCastleResource): Decrypt[F] = new Decrypt[F] {
+    import scala.jdk.CollectionConverters._
+    private val fingerprintCalculator = new JcaKeyFingerprintCalculator
```

Comment:
I just got bit by this recently, but I recommend taking a very defensive posture with this BouncyCastle stuff. I had a `val` in an `object` (that as soon as this happened I just 😵‍💫 was floored) that shouldn't have been and many calls in parallel to said `val` managed to mangle a couple hundred signatures. Not saying that would happen here but better to be cautious I think!

Reply:
Yes exactly, it was because of thread safety issues. In my case it was obvious in hind sight that you wouldn't want to re-use a signature generator because it has state that needs to update for every chunk... 

Why I bring it up here is that it's a code smell that is currently front-of-mind, that's all. This is probably fine with the `JcaKeyFingerprintCalculator`, but how much worse would making this a `def` be?

Reply:
Good call. I ended up wrapping the whole thing in an effect, and also did the same with the creation of the various Java `Iterator`s by BouncyCastle. See ed9ba93fea0bfa8c7db084c775821f2f8927ca38.